waku/v2/protocol/waku_rln_relay/waku_rln_relay_utils.nim

```diff
@@ -795,36 +806,50 @@ proc validateMessage*(rlnPeer: WakuRLNRelay, msg: WakuMessage,
     # accept messages whose epoch is within +-MAX_EPOCH_GAP from the current epoch
     debug "invalid message: epoch gap exceeds a threshold", gap = gap,
         payload = string.fromBytes(msg.payload)
+    waku_rln_invalid_messages_total.inc(labelValues=["invalid_epoch"])
     return MessageValidationResult.Invalid
 
   if not rlnPeer.validateRoot(msg.proof.merkleRoot):
     debug "invalid message: provided root does not belong to acceptable window of roots", provided=msg.proof.merkleRoot, validRoots=rlnPeer.validMerkleRoots
+    waku_rln_invalid_messages_total.inc(labelValues=["invalid_root"])
     return MessageValidationResult.Invalid
 
   # verify the proof
   let
     contentTopicBytes = msg.contentTopic.toBytes
     input = concat(msg.payload, contentTopicBytes)
-    proofVerificationRes = rlnPeer.rlnInstance.proofVerify(input, msg.proof)
+
+  waku_rln_proof_verification_total.inc()
+  waku_rln_proof_verification_duration_seconds.granularTime:
+    let proofVerificationRes = rlnPeer.rlnInstance.proofVerify(input, msg.proof)
 
   if proofVerificationRes.isErr():
+    waku_rln_errors_total.inc(labelValues=["proof_verification"])
     return MessageValidationResult.Invalid
   if not proofVerificationRes.value():
     # invalid proof
     debug "invalid message: invalid proof", payload = string.fromBytes(msg.payload)
+    waku_rln_invalid_messages_total.inc(labelValues=["invalid_proof"])
     return MessageValidationResult.Invalid
 
   # check if double messaging has happened
   let hasDup = rlnPeer.hasDuplicate(msg)
-  if hasDup.isOk and hasDup.value == true:
-    debug "invalid message: message is a spam", payload = string.fromBytes(msg.payload)
+  if hasDup.isErr():
+    waku_rln_errors_total.inc(labelValues=["duplicate_check"])
+  elif hasDup.value == true:
+    debug "invalid message: message is spam", payload = string.fromBytes(msg.payload)
+    waku_rln_spam_messages_total.inc()
     return MessageValidationResult.Spam
 
   # insert the message to the log
   # the result of `updateLog` is discarded because message insertion is guaranteed by the implementation i.e.,
   # it will never error out
   discard rlnPeer.updateLog(msg)
   debug "message is valid", payload = string.fromBytes(msg.payload)
+  let rootLabel = $rlnPeer.validMerkleRoots.find(msg.proof.merkleRoot)
+  waku_rln_valid_messages_total.inc(labelValues=[
+    rootLabel
```

Comment:
What is the cardinality of this label? High cardinality labels have terrible effect on query performance and any prometheus backend existing today. Labels should have a capped number of possible values not exceeding hundreds. Label values that are uncapped can easily explode and cause major performance issues for the whole cluster, as there's no easy ways to clean up specific metrics with specific labels.

What are the example values of `rootLabel`? How man of those values can be? Is there a limit?

Reply:
Examples of root label: 0,1,2..AcceptableRootWindowSize

Ideally we would add a check to make sure that the operator does not set the AcceptableRootWindowSize to a very large number, but due to the nature of the validations, we can expect any value between 0 & 4.

Reply:
Yes, some validation would be good if it is actually possible to set this to any number, since that could easily be used to kill our(or any) metrics backend with an absurd number of labels.

Reply:
Addressed in [0656a0a](https://github.com/status-im/nwaku/pull/1185/commits/0656a0aeb3afc11427e9c33048f76692872a1da6). Not sure if it's the most nim way to do it. 

cc: @LNSD / @jm-clius wdyt?

Reply:
Addressed properly in [ee1ee0a](https://github.com/status-im/nwaku/pull/1185/commits/ee1ee0a31b7835c6d46c84857129d994011edd4d), specifically - 


https://github.com/status-im/nwaku/blob/ee1ee0a31b7835c6d46c84857129d994011edd4d/waku/v2/protocol/waku_rln_relay/waku_rln_relay_metrics.nim#L7-L14 and https://github.com/status-im/nwaku/blob/ee1ee0a31b7835c6d46c84857129d994011edd4d/waku/v2/protocol/waku_rln_relay/waku_rln_relay_metrics.nim#L20-L22

Reply:
@jakubgs, is this solution acceptable?